Fix currency issues

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
```

## What is the current behavior?
- We use english currency symbols in the currency pipe
- By default we round currency values based on the currency format

Issue Number: #20385 #10189


## What is the new behavior?
- We use locale currency symbols
- By default we round currencies based on the number of digits available for that currency


## Does this PR introduce a breaking change?
```
[x] Yes
```

mary-poppins commented:
You can preview 4444e5f at https://pr21783-4444e5f.ngbuilds.io/.

mary-poppins commented:
You can preview e210e4b at https://pr21783-e210e4b.ngbuilds.io/.

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -10,106 +10,48 @@
+];
+
+/** @internal */
+export const CURRENCY_EN: (string | undefined)[][] = [
```
- is the added number related to this PR ?
- the type should be more precise `[string|undefined, string|undefined]` or `[string|undefined, string|undefined, number]`

vicb replied:
`(string | undefined)[]` (more precisely the new one) should be an internal type (something like `CurrencyNameInfo`)

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -10,106 +10,48 @@
-  'XPF': ['CFPF'],
-  'ZAR': [, 'R'],
-  'ZMW': [, 'ZK'],
+export const CURRENCY_CODES = [
```
Seems like we could keep the map here and have the locale files only contain a map of the diff.
I don't think it would make a big difference size wise (and might even be a little smaller) and it would be much easier to understand/debug/extend.

vicb replied:
also should be `Object.create(null, {...})`

ocombe replied:
I tried both, and there was a small difference of about 0.1kb minified in term of size for the fr locale. But I agree that it would be easier to understand

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -24,16 +24,22 @@ export type FormatNumberRes = {
   error?: string
 };
 
+export interface FormatNumberOptions {
```
missing one liner doc.
when is this used / why (w/o blabla)

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -145,17 +151,17 @@ export function formatNumber(
     formattedText = pattern.posPre + formattedText + pattern.posSuf;
   }
 
-  if (style === NumberFormatStyle.Currency && currency !== null) {
```
_note: could this be null ? (doesn't seem so looking at the type)_

ocombe replied:
not with the new "opt" parameter, it can be undefined now

vicb commented on packages/common/src/i18n/locale_data.ts:
```diff
@@ -27,6 +29,7 @@ export function registerLocaleData(data: any, localeId?: string | any, extraData
   localeId = localeId.toLowerCase().replace(/_/g, '-');
 
   LOCALE_DATA[localeId] = data;
+  delete CURRENCIES[localeId];
```
why ?

ocombe replied:
we keep the computed data in memory for performance reasons, but if someone wants to overwrite the locale data, it should replace what we have in memory for that locale

vicb commented on packages/common/src/i18n/locale_data.ts:
```diff
@@ -54,15 +59,30 @@ export const enum LocaleDataIndex {
+export const enum CurrencyIndex {
+  Symbol = 0,
+  SymbolNarrow,
+  Rounding
```
- `Rounding` should no be here ?
- `CurrencyIndex` or the comment does not really help to understand here. Adding a tupe type as suggested earlier would help more.

vicb commented on packages/common/src/i18n/locale_data.ts:
```diff
@@ -54,15 +59,30 @@ export const enum LocaleDataIndex {
 /**
  * Index of each type of locale data from the extra locale data array
+ *
+ * @experimental i18n support is experimental.
```
why is this marked as experimental ? is that exported ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -391,6 +391,25 @@ export function getLocaleCurrencyName(locale: string): string|null {
+/**
+ * Returns the currency values for the locale
+ */
+function getLocaleCurrencies(locale: string): {[code: string]: (string | undefined)[]} {
```
I am not sure what `currency values` or `Currencies` mean here

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -391,6 +391,25 @@ export function getLocaleCurrencyName(locale: string): string|null {
+ * Returns the currency values for the locale
+ */
+function getLocaleCurrencies(locale: string): {[code: string]: (string | undefined)[]} {
+  if(CURRENCIES[locale]) {
```
seems like this is missing format ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -391,6 +391,25 @@ export function getLocaleCurrencyName(locale: string): string|null {
+  }
+  const data = findLocaleData(locale);
+  const localeCurrencies: {[code: string]: (string | undefined)[]} = data[LocaleDataIndex.Currencies];
+  const currenciesDedup: (string | undefined)[][] = data[LocaleDataIndex.CurrenciesDedup];
```
rename to `currenciesDiffEn` ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -391,6 +391,25 @@ export function getLocaleCurrencyName(locale: string): string|null {
+    return CURRENCIES[locale];
+  }
+  const data = findLocaleData(locale);
+  const localeCurrencies: {[code: string]: (string | undefined)[]} = data[LocaleDataIndex.Currencies];
```
not sure what this is.
Please add (short and efficient) comments

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -526,18 +545,19 @@ export function findLocaleData(locale: string): any {
-  const currency = CURRENCIES[code] || [];
-  const symbolNarrow = currency[1];
+export function getCurrencySymbol(code: string, format: 'wide' | 'narrow', locale: string = 'en'): string {
+  const currency = getLocaleCurrencies(locale)[code] || [];
```
the `[]` here is not consistent with the rest of the code (missing the number in the last position) - a tupe type would help here again

vicb commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -143,10 +144,14 @@ export class CurrencyPipe implements PipeTransform {
+      if (display === 'symbol' || display === 'symbol-narrow') {
+        currency = getCurrencySymbol(currency, display === 'symbol' ? 'wide' : 'narrow', locale);
+      } else {
+        currency = display;
```
could this be a boolean ?

ocombe replied:
Not at this point, we test for the boolean value first and replace it with `symbol` or `code` depending on the value.

vicb replied:
Please add `let currency **: string**` to make it clearer

vicb commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -15,10 +15,17 @@ export declare class AsyncPipe implements OnDestroy, PipeTransform {
 }
 
+/** @experimental */
+export declare const enum CurrencyIndex {
```
Is this expected to be public ?

vicb commented on tools/public_api_guard/common/common.d.ts:
```diff
@@ -177,6 +191,31 @@ export declare class JsonPipe implements PipeTransform {
 }
 
+/** @experimental */
+export declare const enum LocaleDataIndex {
```
should not be public ?

vicb commented on packages/common/locales/af-NA.ts:
```diff
@@ -51,6 +51,20 @@ export default [
   'R',
   'Suid-Afrikaanse rand',
   [
+    ,
```
Why ?

ocombe committed:
fix(common): add locale currency values

we now use locale currency symbols, since they may be different in each locale (we were only using english data previously)

Fixes #20385

mary-poppins commented:
You can preview d530a44 at https://pr21783-d530a44.ngbuilds.io/.

mary-poppins commented:
You can preview f9ac110 at https://pr21783-f9ac110.ngbuilds.io/.

vicb commented on packages/common/src/common.ts:
```diff
@@ -13,7 +13,7 @@
  */
 export * from './location/index';
```
Probably the commit message would be more clear as: "use localized currency symbols and names" ?

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -9,49 +9,142 @@
 /** @internal */
-export const CURRENCIES: {[locale: string]: {[code: string]: (string | number | undefined)[]}} = {
-  'en': {}
+export const CURRENCIES: {[code: string]: (string | number | undefined)[]} = {
```
couldn't you use a tuple type ?

vicb commented:
Commits are mixed. I can not review that PR.
Either:
- Separate the commits as (change1, [change1 update1, ...], change 1 generated data, change 2, [change2 update1, ...], change 2 data, ...)
- Open a new PR with the first change only and once it's ok rebase the second change on top of that.

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -9,107 +9,142 @@
+
 /** @internal */
-export const CURRENCIES: {[code: string]: (string | undefined)[]} = {
+export const CURRENCIES: {[code: string]: (string | number | undefined)[]} = {
```
tuple type ?

ocombe replied:
Since the size of each array is different, the tuple type would be:

vicb replied:
yes, this is the more correct version the the type. Defined a type alias an use it everywhere.

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -9,107 +9,142 @@
+  'ADP': [, , 0],
+  'AFN': [, , 0],
+  'ALL': [, , 0],
+  'AMD': [, , 0],
```
why do we have more currencies now ?

ocombe replied:
because when a currency code has no particular symbol (symbol == currency code), we don't add it to the list of currencies because it's unnecessary
but when the digit value is specific (!= 2), we need to add it to the list

vicb replied:
👍

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -24,14 +24,27 @@ export type FormatNumberRes = {
 
+/**
+ * Optional parameters for the function formatNumber
+ * Options are:
```
move the comment on the props

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -24,14 +24,27 @@ export type FormatNumberRes = {
 export function formatNumber(
-    value: number | string, locale: string, style: NumberFormatStyle, digitsInfo?: string | null,
-    currency: string | null = null): FormatNumberRes {
+    value: number | string, locale: string, style: NumberFormatStyle,
```
you should be able to add different signatures:
```
signature1(style: Currency, opt: CurrencyOpt)
signature2(style: Number, opt: NumberOpt)
```
so that you dont' have to use `!` in 
```
+    if (style === NumberFormatStyle.Currency) {
 +      minFraction = getCurrencyDigits(opt.currencyCode !);
 +      maxFraction = minFraction;
 +    }
```

ocombe replied:
would the function be defined like that:

vicb replied:
Yes this is how it would be defined.
The first 3 signatures are for type checking at call sites only.
In the function the types will be as defined on the last line -> you might need to check for a field or cast.
Makes sense ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -391,6 +391,23 @@ export function getLocaleCurrencyName(locale: string): string|null {
+ * Returns the currency values for the locale
+ */
+function getLocaleCurrencies(locale: string):
+    {[code: string]: [string | undefined, string | undefined]} {
```
tuple type ?

ocombe replied:
it is a tuple type, unless you have a different definition for what a tuple type is?

vicb replied:
Should be fixed

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -9,107 +9,142 @@
 
+
 /** @internal */
-export const CURRENCIES: {[code: string]: (string | undefined)[]} = {
```
`CURRENCIES_EN` ? (or at least add a comment)

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -391,6 +391,23 @@ export function getLocaleCurrencyName(locale: string): string|null {
+  const data = findLocaleData(locale);
+  const localeCurrencies = data[LocaleDataIndex.Currencies];
+  const codes = Object.keys(CURRENCIES);
+  codes.forEach((code) => {
```
there is no need to copy (duplicate) existing info that are read-only.
As this should only be used through an API the `locale || en` should be done on read.

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -526,18 +543,34 @@ export function findLocaleData(locale: string): any {
+}
+
+/**
+ * Returns the currency decimal digits for the given currency code.
```
*number of* decimal digits ?

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -526,18 +543,34 @@ export function findLocaleData(locale: string): any {
+export function getCurrencyDigits(code: string): number {
+  const currency = CURRENCIES[code];
+  if (currency && typeof currency[CurrencyIndex.Digits] === 'number') {
+    return <number>currency[CurrencyIndex.Digits];
```
`<number>` is deprecated, use `as number` but it should not be required with a tuple type.

ocombe replied:
It is required because the value can be undefined, and the typeof check doesn't work on non-locale variables.
I've changed the code to use a locale variable so that we don't need to type cast as number

vicb replied:
I understand. 
Here, I am only saying that the `<>` notation is deprecated. `as` should be used instead when casting is needed.

vicb commented on packages/common/src/i18n/locale_data_api.ts:
```diff
@@ -526,18 +543,34 @@ export function findLocaleData(locale: string): any {
+  if (currency && typeof currency[CurrencyIndex.Digits] === 'number') {
+    return <number>currency[CurrencyIndex.Digits];
+  }
+  return 2;
```
2 -> const

vicb commented on packages/common/test/pipes/number_pipe_spec.ts:
```diff
@@ -125,7 +125,13 @@ import {beforeEach, describe, expect, it} from '@angular/core/testing/src/testin
           expect(pipe.transform(5.1234, 'CAD', 'symbol-narrow', '5.2-2')).toEqual('$00,005.12');
           expect(pipe.transform(5.1234, 'CAD', 'symbol-narrow', '5.2-2', 'fr'))
               .toEqual('00 005,12 $');
+          expect(pipe.transform(5, 'USD', 'symbol', '', 'fr')).toEqual('5,00 $US');
```
Could you split this test in multiple test with a narrower description of what they test

vicb commented on packages/common/test/pipes/number_pipe_spec.ts:
```diff
@@ -125,7 +125,13 @@ import {beforeEach, describe, expect, it} from '@angular/core/testing/src/testin
               .toEqual('00 005,12 $');
+          expect(pipe.transform(5, 'USD', 'symbol', '', 'fr')).toEqual('5,00 $US');
           expect(pipe.transform(5.1234, 'FAKE', 'symbol')).toEqual('FAKE5.12');
+          expect(pipe.transform(5.1234, 'USD', 'FAKE')).toEqual('FAKE5.12');
```
May be we should have a test using:
- a currency code that has a non default number of digit (ie <COD>)
- a custom name "custom_<COD>"

and make sure this work.

I am not sure we have a test covering this (it should not use the default)

vicb reviewed (changes requested):
See inline comments
Please split the commits by changes (will require an other review)

mary-poppins commented:
You can preview 624dbcd at https://pr21783-624dbcd.ngbuilds.io/.

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -11,7 +11,10 @@
-export const CURRENCIES: {[code: string]: (string | number | undefined)[]} = {
+export const CURRENCY_EN: {
+  [code: string]:
+      [string] | [string | undefined, string] | [string | undefined, string | undefined, number]
```
Pls define a type alias an re-use it everywhere needed.

vicb commented on packages/common/src/i18n/currencies.ts:
```diff
@@ -11,7 +11,10 @@
 
 /** @internal */
-export const CURRENCIES: {[code: string]: (string | number | undefined)[]} = {
+export const CURRENCY_EN: {
```
CURRENCIES_EN ?

vicb commented on packages/common/test/i18n/locale_data_api_spec.ts:
```diff
@@ -76,10 +76,10 @@ import {findLocaleData, getCurrencySymbol, getLocaleDateFormat, FormatWidth, get
 
     describe('getCurrencyDigits', () => {
```
That's why it's better to describe with words instead of fn names :)

vicb commented on packages/common/test/pipes/number_pipe_spec.ts:
```diff
@@ -126,10 +126,21 @@ import {beforeEach, describe, expect, it} from '@angular/core/testing/src/testin
+
+        it('should support any currency code name', () => {
+          // currency code is unknown, default formatting options will be used
           expect(pipe.transform(5.1234, 'FAKE', 'symbol')).toEqual('FAKE5.12');
```
nit/suggestion: tests are sometime helpful to document the code. In this sense `FAKE` and `symbol` could be replaced with more descriptive name. ie `pipe.transform(5.1234, 'unexisting_ISO_code', 'whatever-value'))`

or `pipe.transform(5.1234, 'USD', 'custom_name')`

ocombe committed:
fix(common): regenerate i18n locale data files

ocombe committed:
fix(common): round currencies based on decimal digits in `CurrencyPipe`

By default, we now round currencies based on the number of decimal digits available for that currency instead of using the rouding defined in the number formats.
More info about that can be found in http://www.unicode.org/cldr/charts/latest/supplemental/detailed_territory_currency_information.html#format_info

Fixes #10189

vicb commented on packages/common/src/i18n/locale_data.ts:
```diff
@@ -66,3 +67,8 @@ export const enum ExtraLocaleDataIndex {
 }
+
+/**
+ * Index of each value in currency data
```
Link to the type you have defined

mary-poppins commented:
You can preview b289fbb at https://pr21783-b289fbb.ngbuilds.io/.

vicb commented on packages/common/test/i18n/locale_data_api_spec.ts:
```diff
@@ -54,16 +56,21 @@ import {findLocaleData, getCurrencySymbol, getLocaleDateFormat, FormatWidth} fro
     });
 
-    describe('getCurrencySymbolElseCode', () => {
+    describe('getCurrencySymbol', () => {
```
use words, not function name

mary-poppins commented:
You can preview f2dd55d at https://pr21783-f2dd55d.ngbuilds.io/.

mary-poppins commented:
You can preview 7002af9 at https://pr21783-7002af9.ngbuilds.io/.

mary-poppins commented:
You can preview da9de45 at https://pr21783-da9de45.ngbuilds.io/.

vicb commented:
g3 patch = cl/185053613
